### PR TITLE
Automatically shuffle train/valid/test inds in megatron

### PR DESCRIPTION
**Pull request by nightingal3**

### CODE REVIEW

File: pretrain_gpt.py [hunk 1]
```diff
@@ -325,7 +325,9 @@ def train_valid_test_datasets_provider(train_val_test_num_samples):
         valid_data_prefix=args.valid_data_path,
         multiple_valid_sets=args.multiple_valid_sets,
         test_data_prefix=args.test_data_path,
-        data_cache_path=args.data_cache_path)
+        data_cache_path=args.data_cache_path,
+        shuffle_docs_before_split=args.shuffle_docs_before_split,
```
Comment (nightingal3) [inline pretrain_gpt.py:329]: new argument added here

File: megatron/data/gpt_dataset.py [hunk 2]
```diff
@@ -104,87 +141,178 @@ def build_train_valid_test_datasets(data_prefix, data_impl, splits_string,
...
         if splits[index + 1] > splits[index]:
-            documents = np.arange(start=splits[index], stop=splits[index + 1],
-                                  step=1, dtype=np.int32)
-            dataset = GPTDataset(name, data_prefix, documents, indexed_dataset,
-                                 splits_string,
-                                 train_valid_test_num_samples[index],
-                                 seq_length, seed,
-                                 return_doc_ids,
-                                 data_cache_path=data_cache_path)
+            if shuffle_docs_before_split:
```
Comment (nightingal3) [inline megatron/data/gpt_dataset.py:231]: change here: shuffle docs to a predictable split in the same split ratios as the original dataset e.g. 98:1:1

File: megatron/data/gpt_dataset.py [hunk 3]
```diff
@@ -104,87 +141,178 @@ def build_train_valid_test_datasets(data_prefix, data_impl, splits_string,
...
+                return_doc_ids,
+                data_cache_path=data_cache_path,
+                shuffle_docs_before_split=shuffle_docs_before_split,
+            )
         return dataset
 
-    train_dataset = build_dataset(0, 'train')
-    valid_dataset = build_dataset(1, 'valid')
-    test_dataset = build_dataset(2, 'test')
+    def doc_idx_to_split(index, seed, train_split_ratios):
```
Comment (nightingal3) [inline megatron/data/gpt_dataset.py:258]: change: new function to handle the top-level splits

File: megatron/data/gpt_dataset.py [hunk 4]
```diff
@@ -195,88 +323,130 @@ def build_dataset(dataset_name, data_prefix, data_impl,
...
+        name,
+        data_prefix,
+        documents,
+        indexed_dataset,
+        splits_string,
+        num_samples,
+        seq_length,
+        seed,
+        return_doc_ids=False,
+        shuffle_docs_before_split=False,
```
Comment (nightingal3) [inline megatron/data/gpt_dataset.py:421]: change: pass arg through to GPTDataset

File: megatron/data/gpt_dataset.py [hunk 5]
```diff
@@ -354,13 +540,15 @@ def _build_index_mappings(name, data_prefix, documents, sizes,
     desc += f"Sequence length {seq_length}\n"
     desc += f"Random seed {seed}\n"
     desc += f"Split {splits_string}\n"
-    desc_hash = hashlib.md5(desc.encode('utf-8')).hexdigest()
+    desc += f"Shuffle original documents {shuffle_orig_docs}\n"
```
Comment (nightingal3) [inline megatron/data/gpt_dataset.py:543]: change: add whether the documents were shuffled at the top level to the description hash. When searching for files, whether the top level assignments were shuffled will also be matched when megatron is deciding whether to construct the dataset or load in existing indices

**Review by CoderPat — APPROVED** [review]
LGTM (the diff makes it hard to parse, but looked at the places you pointed and it looked good)
(just ensure it runs)
